text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+- 2022-01-17
+- status: proposal
+- PR: https://github.com/googlefonts/oxidize/pull/3
+
+## A foundation for building font tooling in Rust
+
+
+# Overview
+
+We propose a Rust library (crate) for zero-allocation access and direct manipulation
+of font tables and types, leveraging Rust’s procedural macro system. This
+library is intended to be an opinion-free transparent representation of the types
+in the specification, to serve as a foundational component for anything that
+needs to read or write font files. An additional goal is to provide a common set
+of types to represent common data formats (tags, glyph identifiers, etc) that
+can be used throughout the broader ecosystem.
+
+
+# Motivation
+
+This is intended to answer some of the questions outlined in this repository's
+README. In particular, we aim to determine whether the following goals are feasible:
+
+- using macros to describe the shape of tables
+-copy generating both zero-allocation read-only table definitions as well as owned,
+  mutable definitions (for compilation)
+- suitable for prototyping & extensibility, i.e. it should be easy to add a new
+  table type for a project without needing to patch the core crate or maintain a
+  fork.
+
+# Design
+
+The particulars of the design will be determined through this exploration, but a
+few core ideas are clear. For the read-only case, the design is largely borrowed
+directly from [pinot][], with the main difference being using macros for code
+generation, and that approach has been successful for [fonttools-rs][].
```

Comment:
QQ, what is unique to the fonttools-rs codegen? I'm under the impression macros are a widely used and well known capability in Rust?

Reply:
Not unique per se, but it is a demonstration that using macros to generate code for these sorts of data structures (which have various weird features like the offset tables and heterogeneous collections) should work.

Reply:
I think the most important thing to learn from fonttools-rs/otspec is the situations where macro-based parsing *hasn't* been successful.

We all agree that the easy cases are easy. It's the hard cases that stress the techniques. OpenType subtables are often badly behaved: many tables are not self-contained and require data from other tables to correctly parse them (maxp/loca/glyf being one example); others have internal dependencies in how they are parsed. ValueRecords show both of these dysfunctions.

It would be very valuable to look at the bits of fonttools-rs which don't use macros, and then compare those with how harfbuzz parses those subtables.